### PR TITLE
multi-selection drag & drop + re-open last selection

### PR DESCRIPTION
**Pull request by mrbuds**
demo: https://i.imgur.com/LIrgzhW.gif

### CODE REVIEW

**Comment by emptyrivers** [comment]
I don't really like how all of the buttons lift up like that. I could see it getting a bit hairy if e.g. you select a couple dozen display buttons. I think having a dummy display that just attaches to the cursor might be better.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 1]
```diff
@@ -268,7 +268,8 @@ local Actions = {
           group.callbacks.UpdateExpandButton();
           group:ReloadTooltip()
         else
-          error("Calling 'Group' with invalid groupId. Reload your UI to fix the display list.")
+          WeakAuras.Add(source.data)
+          -- error("Calling 'Group' with invalid groupId. Reload your UI to fix the display list.")
```
Comment (emptyrivers) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Commented code that doesn't look likely to ever be uncommented

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 2]
```diff
@@ -748,8 +749,10 @@ local methods = {
 
     function self.callbacks.OnDragStart()
       if WeakAuras.IsImporting() or self:IsGroup() then return end;
-      WeakAuras.PickDisplay(data.id)
-      WeakAuras.SetDragging(data)
+      if not WeakAuras.tempGroup.controlledChildren then
```
Comment (emptyrivers) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: As far as I know, tempGroup always has controlled children.

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 3]
```diff
@@ -4075,8 +4075,19 @@ end
 
 function WeakAuras.SetDragging(data, drop)
   WeakAuras_DropDownMenu:Hide()
-  for id, button in pairs(displayButtons) do
-    button:SetDragging(data, drop)
+  if (frame.pickedDisplay == tempGroup and #tempGroup.controlledChildren > 0) then
+    local datas = {}
```
Comment (emptyrivers) [inline WeakAurasOptions/WeakAurasOptions.lua]: the plural of data is still data

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 2 above)
Comment (mrbuds) [reply WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: indeed, good catch !

**Comment by mrbuds** [comment]
I though about your idea of a dummy display, and i think a way to do that would be to only show a subset of the selection and add a transparency mask on the .. mhm .. let's say 3rd item
multi-selection subset
I made this with gimp and i don't know if it is possible with the api

**Comment by mrbuds** [comment]
i tried to implement the mock-up, best i could get for now is this
multi dragging

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 4]
```diff
@@ -1056,20 +1061,40 @@ local methods = {
       self:Enable();
     end
   end,
-  ["SetDragging"] = function(self, data, drop)
+  ["SetDragging"] = function(self, data, drop, displayButtons)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: This is the wrong place to implement most of this code.

This is the "SetDraggin" method of a single button. And that one should not care about other buttons.

You should move most of this code to WeakAuras.SetDragging

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 5]
```diff
@@ -1088,37 +1113,97 @@ local methods = {
...
+        if (self.dragging) then
+          self:Drop(true)
+        end
       end
     end
   end,
   ["ShowTooltip"] = function(self)
   end,
   ["Drag"] = function(self)
-    local uiscale, _, y = UIParent:GetScale(), GetCursorPosition()
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Same here, this is the Drag method of a button. It should not know anything about other buttons.

So the whole logic should be in SetDragging

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 6]
```diff
@@ -4075,8 +4076,19 @@ end
...
-  for id, button in pairs(displayButtons) do
-    button:SetDragging(data, drop)
+  if (frame.pickedDisplay == tempGroup and #tempGroup.controlledChildren > 0) then
+    local children = {}
+    for index, childId in pairs(tempGroup.controlledChildren) do
+      local data = WeakAuras.GetData(childId);
+      children[data.id] = index
+    end
+    for id, button in pairs(displayButtons) do
+      button:SetDragging(data and children, drop, displayButtons);
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: Here for each button you drag, you call button:SetDragging, which will then iterate over all displayButtons, essentially each call to button:SetDragging doing the same work.

Comment (mrbuds) [reply WeakAurasOptions/WeakAurasOptions.lua]: button:SetDragging iterate over all displayButtons only once `(displayButtons and data[self.data.id] == 1)`

**Comment by mrbuds** [comment]
Thanks for your comments, i moved the iteration on selected buttons on WeakAuras.SetDragging, it made the code much easier.
Also, i was moving buttons out of screen because Hide() removed the drop event if i didnt select the first item from the list, this issue is now solved by using the selected button instead of the 1st from the list.

The last issue i see is with how i handle the temporary "WA" icon, if the button i'm dragging already have an icon, it will have both icon at the same time while dragging

**Comment by mrbuds** [comment]
Last commit fix a regression when closing WeakAurasOptions with more than one button selected

It also implement a suggestion from @Stanzilla on discord to re-select previous button(s) when re-opening WeakAurasOptions

**Comment by Stanzilla** [comment]
merged as https://github.com/WeakAuras/WeakAuras2/pull/744

**Pull request closed without merging**